### PR TITLE
Use std::vector instead of array, remove base/tl/array.h, algorithm.h, allocator.h, range.h

### PR DESCRIPTION
**Pull request by Robyt3**
Replaces all usages of `array` with `std::vector`.

I adjusted variable names of variables I changed to use the `v` prefix. Not so in the editor however, as there are already many changes due to clang-tidy enforcing the use of for-each loops.

This allows us to remove all remaining `base/tl` headers except `threading.h`.

Clang-tidy now finds `clang-analyzer-cplusplus.NewDelete` (Use of memory after it is freed), which is also fixed here, though it appears to be a false-positive.

This last remaining usages of `goto` are also removed.

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Review by def- — APPROVED** [review]
lgtm, getting closer to standard c++ :)

**Comment by Learath2** [comment]
Sue me but that `goto` was perfectly valid and readable there :(
I like the rest LGTM

**Comment by def-** [comment]
I don't care either way about the gotos

**Comment by C0D3D3V** [comment]
grafik

code is a lot cleaner I like it :) the goto would not make it better here

**Comment by Learath2** [comment]
> grafik
Does the right one actually look better to you? When that go to not only makes it more concise but is also extremely readable? If so I say you've been brainwashed by Djikstra and we have nothing to discuss. You'll never change your mind.

On a more serious note, I don't actually care one way or the other even though the left one is infinitely prettier, rest of the CS world seems to have moved on without me. Can be merged as is

**Comment by def-** [comment]
My proposal:
```
const bool SoundUsed = std::any_of(m_Map.m_lGroups.cbegin(), m_Map.m_lGroups.cend(), [i](const auto &pGroup) {
	return std::any_of(pGroup->m_lLayers.cbegin(), pGroup->m_lLayers.cend(), [i](const auto &pLayer) {
		return static_cast<CLayerSounds *>(pLayer)->m_Sound == i;
	});
});
```

File: src/game/editor/editor.cpp [hunk 1]
```diff
@@ -4268,24 +4250,13 @@ void CEditor::RenderSounds(CUIRect ToolBox, CUIRect View)
...
 
 		int Selected = m_SelectedSound == i;
-		bool SoundUsed = false;
-		for(const auto &pGroup : m_Map.m_lGroups)
-		{
-			for(const auto &pLayer : pGroup->m_lLayers)
-			{
+		const bool SoundUsed = std::any_of(m_Map.m_lGroups.cbegin(), m_Map.m_lGroups.cend(), [i](const auto &pGroup) {
+			return std::any_of(pGroup->m_lLayers.cbegin(), pGroup->m_lLayers.cend(), [i](const auto &pLayer) {
 				if(pLayer->m_Type == LAYERTYPE_SOUNDS)
```
Comment (def-) [inline src/game/editor/editor.cpp:4255]: I forgot that if, thanks for including it :)

**Pull request merged**